File: coci/coci08c4p4/coci08c4p4.cpp
#include <algorithm>
#include <iostream>

int n;
char grid[512][513];
char out[512][513];

struct data {
	int all0, all1, rec;
};

data recurse(int x1, int x2, int y1, int y2) {
	data ret = {0, 0, 262144};
	if (x2 - x1 == 1) {
		out[y1][x1] = grid[y1][x1];
		ret.all0 = grid[y1][x1] == '1';
		ret.all1 = grid[y1][x1] == '0';
		ret.rec = 0;
	} else {
		int size = x2 - x1 >> 1;
		int xmid = x1 + size;
		int ymid = y1 + size;
		data squares[] = {recurse(x1, xmid, y1, ymid), recurse(xmid, x2, y1, ymid), recurse(x1, xmid, ymid, y2), recurse(xmid, x2, ymid, y2)};
		for (data square : squares) {
			ret.all0 += square.all0;
			ret.all1 += square.all1;
		}
		int best[4];
		int perm[] = {0, 1, 2, 3};
		while (std::next_permutation(perm, perm + 4)) {
			int newcost = squares[perm[0]].all0 + squares[perm[1]].all1 + squares[perm[2]].rec + squares[perm[3]].rec;
			if (newcost < ret.rec) {
				ret.rec = newcost;
				for (int i = 0; i < 4; i++) {
					best[i] = perm[i];
				}
			}
		}
		for (int dy = 0; dy < size; dy++) {
			for (int dx = 0; dx < size; dx++) {
				out[y1 + size * (best[0] >> 1) + dy][x1 + size * (best[0] & 1) + dx] = '0';
				out[y1 + size * (best[1] >> 1) + dy][x1 + size * (best[1] & 1) + dx] = '1';
			}
		}
	}
	return ret;
}

int main() {
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		scanf("%s", grid[i]);
	}
	printf("%d\n", recurse(0, n, 0, n).rec);
	for (int i = 0; i < n; i++) {
		printf("%s\n", out[i]);
	}
	return 0;
}

File: coci/coci08c4p4/coci08c4p4.py
def recurse(x1, x2, y1, y2):
	if x2 - x1 == 1:
		out[y1][x1] = grid[y1][x1]
		return grid[y1][x1] > "0", grid[y1][x1] < "1", 0
	size = x2 - x1 >> 1
	xmid = x1 + size
	ymid = y1 + size
	squares = recurse(x1, xmid, y1, ymid), recurse(xmid, x2, y1, ymid), recurse(x1, xmid, ymid, y2), recurse(xmid, x2, ymid, y2)
	all0 = sum(square[0] for square in squares)
	all1 = sum(square[1] for square in squares)
	cost = 262144
	for a, b, c, d in __import__("itertools").permutations(xrange(4)):
		newcost = squares[a][0] + squares[b][1] + squares[c][2] + squares[d][2]
		if newcost < cost:
			cost = newcost
			best = a, b, c, d
	for dy in xrange(size):
		for dx in xrange(size):
			out[y1 + size * (best[0] >> 1) + dy][x1 + size * (best[0] & 1) + dx] = "0"
			out[y1 + size * (best[1] >> 1) + dy][x1 + size * (best[1] & 1) + dx] = "1"
	return all0, all1, cost


input = __import__("sys").stdin.readline
n = int(input())
grid = [input() for _ in xrange(n)]
out = [[""] * n for _ in xrange(n)]
print(recurse(0, n, 0, n)[2])
print("\n".join("".join(out[i]) for i in xrange(n)))